File: lint-arwaky/archive/taxonomy/symbol_collection_vo.py
from pydantic import BaseModel, ConfigDict, Field, model_validator
from .symbol_name_vo import SymbolName


class SymbolNameList(BaseModel):
    """Collection of symbol names."""

    model_config = ConfigDict(frozen=True)

    @model_validator(mode="before")
    @classmethod
    def coerce_from_list(cls, data: object) -> object:
        if isinstance(data, list):
            return {"values": data}
        return data

    values: list[SymbolName] = Field(default_factory=list)

    def __iter__(self):
        return iter(self.values)

    def __len__(self) -> int:
        return len(self.values)


class ImportNameList(BaseModel):
    """Collection of imported module/symbol names."""

    model_config = ConfigDict(frozen=True)

    @model_validator(mode="before")
    @classmethod
    def coerce_from_list(cls, data: object) -> object:
        if isinstance(data, list):
            return {"values": data}
        return data

    values: list[SymbolName] = Field(default_factory=list)

    def __len__(self) -> int:
        return len(self.values)


class PrimitiveTypeList(BaseModel):
    """List of primitive type names."""

    model_config = ConfigDict(frozen=True)

    @model_validator(mode="before")
    @classmethod
    def coerce_from_list(cls, data: object) -> object:
        if isinstance(data, list):
            return {"values": data}
        return data

    values: list[SymbolName] = Field(default_factory=list)

    def __contains__(self, item: str) -> bool:
        return any((v.value == item for v in self.values))

    def __iter__(self):
        return iter(self.values)

    def __len__(self) -> int:
        return len(self.values)


class CallChainList(BaseModel):
    """List of call site strings (or symbol locations)."""

    model_config = ConfigDict(frozen=True)

    @model_validator(mode="before")
    @classmethod
    def coerce_from_list(cls, data: object) -> object:
        if isinstance(data, list):
            return {"values": data}
        return data

    values: list[SymbolName] = Field(default_factory=list)

    def __iter__(self):
        return iter(self.values)

    def __len__(self) -> int:
        return len(self.values)


CORE_PRIMITIVE_TYPES = ["str", "int", "float"]

PRIMITIVE_TYPE_LIST = PrimitiveTypeList(
    values=[SymbolName(value=t) for t in CORE_PRIMITIVE_TYPES]
)

File: lint-arwaky/src-rust/taxonomy/symbol_collection_vo.rs
use serde::{Serialize, Deserialize};
use super::*;

pub const CORE_PRIMITIVE_TYPES: &[&str] = &["str", "int", "float"];

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SymbolNameList {
    #[serde(default)]
    pub values: Vec<SymbolName>,
}

impl SymbolNameList {
    pub fn new() -> Self { Self { values: Vec::new() } }
    pub fn push(&mut self, item: SymbolName) { self.values.push(item); }
    pub fn len(&self) -> usize { self.values.len() }
    pub fn is_empty(&self) -> bool { self.values.is_empty() }
    pub fn iter(&self) -> std::slice::Iter<'_, SymbolName> { self.values.iter() }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ImportNameList {
    #[serde(default)]
    pub values: Vec<SymbolName>,
}

impl ImportNameList {
    pub fn new() -> Self { Self { values: Vec::new() } }
    pub fn len(&self) -> usize { self.values.len() }
    pub fn is_empty(&self) -> bool { self.values.is_empty() }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PrimitiveTypeList {
    #[serde(default)]
    pub values: Vec<SymbolName>,
}

impl PrimitiveTypeList {
    pub fn new() -> Self { Self { values: Vec::new() } }
    pub fn contains(&self, item: &str) -> bool {
        self.values.iter().any(|v| v.value == item)
    }
    pub fn len(&self) -> usize { self.values.len() }
    pub fn is_empty(&self) -> bool { self.values.is_empty() }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CallChainList {
    #[serde(default)]
    pub values: Vec<SymbolName>,
}

impl CallChainList {
    pub fn new() -> Self { Self { values: Vec::new() } }
    pub fn len(&self) -> usize { self.values.len() }
    pub fn is_empty(&self) -> bool { self.values.is_empty() }
}

pub fn primitive_type_list() -> PrimitiveTypeList {
    PrimitiveTypeList {
        values: CORE_PRIMITIVE_TYPES.iter().map(|s| SymbolName::new(*s)).collect(),
    }
}
